src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,288 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "pal_types.h"
+#include "pal_utilities.h"
+#include "pal_gssapi.h"
+
+#if HAVE_GSSFW_HEADERS
+#include <GSS/GSS.h>
+#else
+#include <gssapi/gssapi_ext.h>
+#include <gssapi/gssapi_krb5.h>
+#endif
+
+#include <assert.h>
+#include <string.h>
+
+static_assert(PAL_GSS_C_DELEG_FLAG == GSS_C_DELEG_FLAG, "");
+static_assert(PAL_GSS_C_MUTUAL_FLAG == GSS_C_MUTUAL_FLAG, "");
+static_assert(PAL_GSS_C_REPLAY_FLAG == GSS_C_REPLAY_FLAG, "");
+static_assert(PAL_GSS_C_SEQUENCE_FLAG == GSS_C_SEQUENCE_FLAG, "");
+static_assert(PAL_GSS_C_CONF_FLAG == GSS_C_CONF_FLAG, "");
+static_assert(PAL_GSS_C_INTEG_FLAG == GSS_C_INTEG_FLAG, "");
+static_assert(PAL_GSS_C_ANON_FLAG == GSS_C_ANON_FLAG, "");
+static_assert(PAL_GSS_C_PROT_READY_FLAG == GSS_C_PROT_READY_FLAG, "");
+static_assert(PAL_GSS_C_TRANS_FLAG == GSS_C_TRANS_FLAG, "");
+static_assert(PAL_GSS_C_DCE_STYLE == GSS_C_DCE_STYLE, "");
+static_assert(PAL_GSS_C_IDENTIFY_FLAG == GSS_C_IDENTIFY_FLAG, "");
+static_assert(PAL_GSS_C_EXTENDED_ERROR_FLAG == GSS_C_EXTENDED_ERROR_FLAG, "");
+static_assert(PAL_GSS_C_DELEG_POLICY_FLAG == GSS_C_DELEG_POLICY_FLAG, "");
+
+static_assert(PAL_GSS_COMPLETE == GSS_S_COMPLETE, "");
+static_assert(PAL_GSS_CONTINUE_NEEDED == GSS_S_CONTINUE_NEEDED, "");
+
+#if !HAVE_GSS_SPNEGO_MECHANISM
+static char gss_mech_value[] = "\x2b\x06\x01\x05\x05\x02"; // Binary representation of SPNEGO Oid (RFC 4178)
+#endif
+
+static uint32_t NetSecurityNative_HandleError(uint32_t majorStatus,
+                                              gss_buffer_t gssBuffer,
+                                              struct PAL_GssBuffer *targetBuffer)
+{
+    assert(targetBuffer != nullptr);
+    assert(gssBuffer != nullptr);
+    if (GSS_ERROR(majorStatus))
+    {
+        assert (gssBuffer->value == nullptr);
+        targetBuffer->length = 0;
+        targetBuffer->data = nullptr;
+        return majorStatus;
+    }
+
+    targetBuffer->length = gssBuffer->length;
```

Comment:
In this (the else) path, do we expect that `gssBuffer->value` is not null?  Is it assert-worthy?

If a nullptr were to come through and everything would be just fine, that's cool; I'm just asking questions.